test/controllers/tasks_controller_test.rb

```diff
@@ -71,56 +70,167 @@
+      }.must_differ "Task.count", 0
+      
+      must_respond_with :redirect
+      must_redirect_to task_path
+      
+      current_task = Task.find_by(id: current_id)
+      expect(current_task.name).must_equal new_task_details[:task][:name]
+      expect(current_task.description).must_equal new_task_details[:task][:description]
+      expect(current_task.completion_date).wont_be_nil
+      expect(current_task.completion_date).must_be_instance_of ActiveSupport::TimeWithZone
+      
     end
-
+    
     it "will redirect to the root page if given an invalid id" do
       # Your code here
+      invalid_id = - 1
+      
+      patch task_path(invalid_id), params: new_task_details
+      
+      must_redirect_to root_path
+    end
+    
+    it "won't update if the params are invalid" do
+      test_id = Task.first.id
+      unchanged_task = Task.find_by(id: test_id)
+      
+      expect {
+        patch task_path(test_id), params: {}
+      }.must_raise
+      
+      updated_task = Task.find_by(id: test_id)
+      expect(updated_task.name).must_equal unchanged_task.name
+      expect(updated_task.description).must_equal unchanged_task.description
+      expect(updated_task.completion_date).must_equal unchanged_task.completion_date
     end
   end
-
+  
   # Complete these tests for Wave 4
   describe "destroy" do
     # Your tests go here
-
+    it "successfully deletes the selected task and redirects back to the list of tasks" do
+      Task.create(name: "Test task!", description: "Testing this task", completion_date: nil)
+      valid_id = Task.find_by(name: "Test task!").id
+      expect {delete task_path(valid_id)}.must_differ "Task.count", -1
+      must_redirect_to tasks_path
+    end
+    
+    it "does not delete any tasks and redirects back to the list of tasks if the given task does not exist" do
+      Task.destroy_all
+      invalid_id = 1
+      
+      expect {delete task_path(invalid_id)}.must_differ "Task.count", 0
+      must_redirect_to tasks_path
+    end
+    
+    it "does not delete any tasks and redirects to the list of tasks if the user tries to delete a task that was already deleted" do
+      Task.create(name: "Another test task!", description: "Testing this task again", completion_date: nil)
+      valid_id = Task.find_by(name: "Another test task!").id
+      delete task_path(valid_id)
+      expect {delete task_path(valid_id)}.must_differ "Task.count", 0
+      must_redirect_to tasks_path
+    end
   end
-
+  
   # Complete for Wave 4
   describe "toggle_complete" do
     # Your tests go here
+    it "succesfully updates the completion date to not be nil if it was previously nil and redirects to index page" do
+      new_task_1 = Task.create(name: "Toggle test task!", description: "Testing the toggle task")
+      expect(new_task_1.completion_date).must_be_nil
+      test_id = new_task_1.id
+      
+      patch toggle_path(test_id)
+      new_task_1.reload
```

Comment:
Nice! `.reload` is a clean way to reload the data, and it looks really clean too! Keep doing this whenever this makes sense!